README
		 fbnews - a RSS fetcher for fluxbox 


fbnews is a RSS/RDF newsfetcher for the window manager fluxbox created in 2004-2005. 
It is currently unmaintained, since don't use fluxbox anymore.

fbnews creates a submenu structure in the fluxbox menu and fills it with the content of 
the RSS sources you specify.  


LICENSE
=========

This program is free software; you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation; either version 2 of the License, or
(at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program; if not, write to the Free Software
Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA

(C) 2004-2005 F. Engel
	



Changelog
Version 0.5.1 (May 2005)
- Fixed bug in doc Makefile (reported by agl)

Version 0.5 (May 2005)
- Fixed bug in config reparsing that lead to crash 
- Html output
- Minor Fixes
Note: Probably the last version released by me

Version 0.4 (Jan 2005)
- News items are arranged as they appear in the rss file
- Subcategory support
- A "Max length" option for site titles

Version 0.3  (Dec 2004)
- Ability to use scripts instead of rss files
- "Once" command line options to not run fbnews as a daemon
- Fixed bug that caused firefox to not startup when clicking a news item
- Real documentation
- added several command line options


Version 0.2  (Oct 2004)
====================
- Now compatible with fluxbox 0.9.10
- Support for non ascii characters
- New, better xml parser


Version 0.1  (Mar 2004)
===================
Second release:
	Thanks to Christopher Bratlien the default browser command
	now opens a new tab in firefox if an instance of it is available.

First release of fbnews.


odict.py
#!/usr/bin/env python
# -*- coding: iso-8859-15 -*-
############################################################################
#    Copyright (C) 2005 by Frithjof Engel
#    frithjof_engel@users.sourceforge.net
#
#    This program is free software; you can redistribute it and#or modify
#    it under the terms of the GNU General Public License as published by
#    the Free Software Foundation; either version 2 of the License, or
#    (at your option) any later version.
#
#    This program is distributed in the hope that it will be useful,
#    but WITHOUT ANY WARRANTY; without even the implied warranty of
#    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#    GNU General Public License for more details.
#
#    You should have received a copy of the GNU General Public License
#    along with this program; if not, write to the
#    Free Software Foundation, Inc.,
#    59 Temple Place - Suite 330, Boston, MA  02111-1307, USA.             
##########################################################################
'''This module implements an ordered dictionary.'''
__revision__ = "0.5"

class odict(dict):
    def __init__(self):
        self._keylist = []
        dict.__init__(self)

    def __delitem__(self, key):
        dict.__delitem__(self, key)
        self._keylist.remove(key)

    def __setitem__(self, key, item):
        dict.__setitem__(self, key, item)
        if key not in self._keylist: self._keylist.append(key)

    def __repr__(self):
        str = u"{"
        for k in self._keylist:
            str += k.__repr__()+": "
            str += self[k].__repr__()+", "

        str += "}"
        return str
            
    def __str__(self):
        return self.__repr__()
            
    def clear(self):
        dict.clear()
        self._keylist = []

    def copy(self):
        d = dict.copy()
        d._keylist = self._keylist[:]
        return d

    def items(self):
        return zip(self._keylist, self.values())

    def keys(self):
        return self._keylist

    def popitem(self):
        try:
            key = self._keylist[-1]
        except IndexError:
            raise KeyError('Dictionary is empty')

        val = self[key]
        del self[key]

        return (key, val)

    def setdefault(self, key, fail = None):
        dict.setdefault(self, key, fail)
        if key not in self._keylist: self._keylist.append(key)

    def update(self, dic):
        dict.update(self, dic)
        for key in dic.keys():
            if key not in self._keylist: self._keylist.append(key)

    def values(self):
        return map(self.get, self._keys)
    
        
if __name__ == '__main__':
    l = odict()
    


fbnews.py
#!/usr/bin/python
# -*- coding: iso-8859-15 -*-
############################################################################
#    Copyright (C) 2004 by Frithjof Engel
#    frithjof_engel@users.sourceforge.net
#
#    This program is free software; you can redistribute it and#or modify
#    it under the terms of the GNU General Public License as published by
#    the Free Software Foundation; either version 2 of the License, or
#    (at your option) any later version.
#
#    This program is distributed in the hope that it will be useful,
#    but WITHOUT ANY WARRANTY; without even the implied warranty of
#    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#    GNU General Public License for more details.
#
#    You should have received a copy of the GNU General Public License
#    along with this program; if not, write to the
#    Free Software Foundation, Inc.,
#    59 Temple Place - Suite 330, Boston, MA  02111-1307, USA.             
##########################################################################
'''This module is the main module of fbnews. It does the actual translation
work to the fluxbox menu format.'''
__revision__ = "0.5"

DEBUG = False
ENCODING = 'iso8859_15'

import urllib, time, os, codecs, sys, getopt, StringIO
import fbnconf
import parser

config = fbnconf.Config()

def fetchRDF(url):
    '''Takes an url and returns a list, with the title as the first element and
    a dictionary of "newitem" : "url"'''

    # Decide the next action based on whether the url is a seperator, script,
    # or real url.
    xmldata = ""
    if url.startswith('Sep:'):
        if DEBUG:
            print 'Sep: '+url
        return url
    elif url.startswith('Script:'):
        if DEBUG:
            print 'Script: ' + url

	scriptout = os.popen(url[url.find(':')+1:].strip())
	url = os.tmpfile()
        url.write(scriptout.read())
        url.seek(0)

    elif DEBUG:
        print 'Rss: '+url

    p = parser.Parser(url)
    title = p.siteTitle()
    dic = p.newsItems()
    return [title, dic]


def toFBMenuFormat(lst):
    '''Converts the list as returned by fetchRDF() to the fluxbox menu file
    format. It is invoked for every category. Returns the string.'''

    if lst[0] == 'S':
        return '[nop] ('+lst[lst.find(':')+1:].strip()+')'
        
    format = '[submenu] ('

    caption = lst[0].replace(')', '\)')
    if config.maxlen() > 0:
    	format += config.siteformat(caption[0:config.maxlen()]) +') {}\n'
    else:
    	format += config.siteformat(caption) + ') {}\n'
    
    for key in lst[1].keys():
        format += u'[exec] (' + key + ') {' \
                  + config.browser(lst[1][key])  + u'}\n'

    format += '[end]\n'
    return format

def toHtmlFormat(lst):
    '''Converts the list to html format.'''

    if lst[0] == 'S':
        return '\n<hr>'

    format = '<h3>'+config.siteformat(lst[0])+'</h3>'
    for key in lst[1].keys():
        format += u'\n<a href='+lst[1][key]+'>'+key+'</a><br>'

    return format

def htmlmain(newsfile):
    '''main loop routine for html format'''
    newsfile.write('<html><title>fbnews (html)</title><body>')

    out = " "
    for cat in config.categories().keys():
        if DEBUG:
            print 'Category: ', cat
        
        if cat != " ":
            out += "\n<h1>"+cat+"</h1>"
                            
        for src in config.categories()[cat]:
            try:
                if out:
                    newsfile.write(out+'\n')
                    data = fetchRDF(src) 
                    if DEBUG: print 'fetched rdf'
                    out = toHtmlFormat(data)
                        
            except:
                config.log('Skipping file: '+src)
                if DEBUG:
                    print 'Skipping file. '+src
                    out = ""
                continue

    newsfile.write('</body></html>')

def fbmain(newsfile):
    '''main loop routine for fbmenu format'''
    out = " "
    newsfile.write("[begin] (fbnews)\n")
    for cat in config.categories().keys():
        if DEBUG:
            print 'Category: ', cat
            
        if cat != " ":
            out += "[submenu] ("+cat+") {}\n"
                            
        for src in config.categories()[cat]:
            try:
                if out:
                    newsfile.write(out+'\n')
                    data = fetchRDF(src) 
                    if DEBUG: print 'fetched rdf'
                    out = toFBMenuFormat(data)
                        
            except:
                config.log('Skipping file: '+src)
                if DEBUG:
                    print 'Skipping file. '+src
                    out = ""
                continue

        if cat != " ":
            out += "\n[end]\n"

    try:
        newsfile.write(out)
    except:
        print out

    newsfile.write(u'[end]\n')


def usage():
    return "Usage: fbnews [options]\nOptions:\n"\
           "  -1, --once        Just fetch feeds once, then exit\n"\
           "  -f rcfile,        Use alternative rcfile as fbnewsrc\n"\
           "  --file rcfile     \n"\
           "  --dump            Dump output to screen\n"\
           "  --html file       Write html to file\n"\
           "  -d, --debug       Be verbose on the screen\n"\
           "  -h, --help        Print this screen\n"

def main():
    '''Main function. Call this to invoke the program.'''
    os.nice(20)

    try:
        optlist, args = getopt.getopt(sys.argv[1:], 'd1hf:',
                                      ['once', 'dump','html=', 'debug', 'help',\
                                       'file='])
    except getopt.GetoptError:
        print usage()
        sys.exit(1)

    Once = False
    Dump = False
    Htmlout = False
    Configfile = ""
    global DEBUG
    for o, v in optlist:
        if o in ('-1', '--once'):
            Once = True
        if o in ('-h', '--help'):
            print usage()
            sys.exit(0)
        if o in ('-f', '--file'):
            Configfile = v
	    config.reparse(v)
        if o in ('-d', '--debug'):
            DEBUG = True
        if o == '--dump':
            Dump = True
        if o == '--html':
            Htmlout = True
            Htmlfile = v
    
    while 1:
	try:
            if Dump:
                newsfile = sys.stdout
            else:
                if not Htmlout:
                    newsfile = codecs.open(config.newsfile(), 'w', ENCODING)
                else:
                    newsfile = codecs.open(Htmlfile, 'w', ENCODING)
            config.log(time.strftime("%H:%M")+': Starting news update...')
	except:
            print 'Could not open newsfile ' + config.newsfile() +\
                  ' for writing'

        if not Htmlout:
            fbmain(newsfile)
        else:
            htmlmain(newsfile)
            
        newsfile.close()

        if Once: sys.exit(0)
	
        config.log('Done fetching data, resting for '+ \
                   str(config.sleep())+ ' minutes...'+"\n")
	
	time.sleep(60 * config.sleep())
        if Configfile:
            config.reparse(Configfile)
        else:
            config.reparse()

if __name__ == '__main__':
    main()


fbnconf.py
#!/usr/bin/env python
############################################################################
#    Copyright (C) 2004-2005 by Frithjof Engel
#    frithjof_engel@users.sourceforge.net
#
#    This program is free software; you can redistribute it and#or modify
#    it under the terms of the GNU General Public License as published by
#    the Free Software Foundation; either version 2 of the License, or
#    (at your option) any later version.
#
#    This program is distributed in the hope that it will be useful,
#    but WITHOUT ANY WARRANTY; without even the implied warranty of
#    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#    GNU General Public License for more details.
#
#    You should have received a copy of the GNU General Public License
#    along with this program; if not, write to the
#    Free Software Foundation, Inc.,
#    59 Temple Place - Suite 330, Boston, MA  02111-1307, USA.             
##########################################################################
'''This module handles all the fbnews stuff that is configurable by 
rc file or by command line. Does not provide writing support.'''
__revision__ = "0.5"

import os.path
import odict

RCFILE = "~/.fluxbox/fbnewsrc"

class Config:
    '''Implements the config class that holds the configuration values.'''
    def __init__(self, filename = RCFILE):
	self.reparse(filename)
        
    def reparse(self, filename=RCFILE):
        '''Does the actual parsing'''
        config = ""
        self.__browser = 'xterm lynx %s'
        self.__curCat = " "
        self.__categories = odict.odict()
	self.__categories[self.__curCat] = []
        self.__siteformat = ''
        self.__newsfile = ""
        self.__logfile = ""
        self.__sleep = 0
        self.__maxLen = -1     
        
        try:
            config = file(os.path.expanduser(filename))
        except IOError:
            print 'Could not read configuration file ' + filename
            pass

        src = 0
        # Get from config file
        for line in config:
            if line.startswith('#') or not line:
                continue
            elif line.startswith('Cat'):
                self.__curCat = line[line.find(':')+1:].strip()
            elif line.startswith('Browser'):
                self.setBrowser(line[line.find(':')+1:].strip())
            elif line.startswith('Newsfile'):
                self.setNewsfile(line[line.find(':')+1:].strip())
            elif line.startswith('Logfile'):
                self.setLogfile(line[line.find(':')+1:].strip())
            elif line.startswith('Siteformat'):
                self.setSiteformat(line[line.find(':')+1:].strip())
            elif line.startswith('Sleep'):
                self.setSleep(int(line[line.find(':')+1:].strip()))
	    elif line.startswith('Maxlen'):
	    	self.setMaxlen(int(line[line.find(':')+1:].strip()))
            elif line.startswith('[Sources]'):
                src = 1
            elif line.startswith('Script:'):
                if not src:
                    self.log('Unrecognized line: '+line)
                    
                if not self.__categories.has_key(self.__curCat):
                        self.__categories[self.__curCat] = []
                self.__categories[self.__curCat].append(line.strip())
	    else:
                line = line.strip()
                if src and len(line):
                    if not self.__categories.has_key(self.__curCat):
                        self.__categories[self.__curCat] = []
                    self.__categories[self.__curCat].append(line)
                else:
                    if len(line):
                        self.log('Unrecognized line: '+line)

    def setBrowser(self, val):
        '''Set the configuration value for browser'''
        self.__browser = val.replace('"','')

    def browser(self, url):
        '''Returns the browser command'''
        return self.__browser.replace('%s', '"'+url+'"')
                              
    def categories(self):
        '''Returns the list of categories'''
        return self.__categories

    def setSiteformat(self, val):
        '''Set the configuration value for the siteformat'''
        self.__siteformat = val.replace('"','')

    def siteformat(self, site):
        '''Returns the siteformat'''
        return self.__siteformat.replace('%s', site)

    def maxlen(self):
    	return self.__maxLen

    def setMaxlen(self, m):
    	self.__maxLen = m
    
    def setNewsfile(self, val):
        '''Set the newsfile name'''
        self.__newsfile = os.path.expanduser(val)

    def newsfile(self):
        '''Returns the newsfilename'''
        return self.__newsfile

    def setLogfile(self, val):
        '''Set the logfile name'''
        self.__logfile = os.path.expanduser(val)
        filehandle = file(self.__logfile,'w') # Clean file
        filehandle.close()

    def log(self, msg):
        '''Returns the logfilename'''
        if not self.__logfile:
            return
        
        filehandle = file(self.__logfile, 'a')
        filehandle.write(msg+'\n')
        filehandle.close()

    def setSleep(self, val):
        '''The the number of minutes to sleep'''
        self.__sleep = val

    def sleep(self):
        '''Returns the number of minutes to sleep'''
        return self.__sleep

if __name__ == '__main__':
    cfg = Config('fbnewsrc')
    for c in cfg.categories():
        print c, ':', cfg.categories()[c]
        print '--------------------'
    


parser.py
#!/usr/bin/env python
# -*- coding: iso-8859-15 -*-
############################################################################
#    Copyright (C) 2004, 2005 by Frithjof Engel
#    frithjof_engel@users.sourceforge.net
#
#    This program is free software; you can redistribute it and/or modify
#    it under the terms of the GNU General Public License as published by
#    the Free Software Foundation; either version 2 of the License, or
#    (at your option) any later version.
#
#    This program is distributed in the hope that it will be useful,
#    but WITHOUT ANY WARRANTY; without even the implied warranty of
#    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#    GNU General Public License for more details.
#
#    You should have received a copy of the GNU General Public License
#    along with this program; if not, write to the
#    Free Software Foundation, Inc.,
#    59 Temple Place - Suite 330, Boston, MA  02111-1307, USA.             
##########################################################################
'''This module contains the rdf parsing code. Example usage:
p = Parser(\'newsfile.rdf\')
print p.siteTitle()
print p.newsItems()
'''

import sys
import odict
from xml.sax import ContentHandler

__revision__ = "0.5"

class Parser:
    '''Public class to instantiate from: Gives interface to the
    data stored in the xml file'''
    def __init__(self, filename):
        from xml.sax import make_parser
        from xml.sax.handler import feature_namespaces
        # Create a parser
        parser = make_parser()

        # Tell the parser we are not interested in XML namespaces
        parser.setFeature(feature_namespaces, 0)

        # Create the handler
        contentHandler = TagHandler()

        # Tell the parser to use our handler
        parser.setContentHandler(contentHandler)

        # Parse the input
        parser.parse(filename)

        self.__newsMap = odict.odict()
        self.__newsList = ()
        for i in contentHandler.itemList:
            title = i[0].replace('\n', '').strip()
            link = i[1].replace('\n', '').strip()
            self.__newsMap[title] = link
                                              
        self.__siteTitle = contentHandler.siteTitle.replace('\n', '').strip()
        self.__siteLink = contentHandler.siteLink.replace('\n', '').strip()
        

#    def addNewsItem(self, title, link):
#        if not title in self.__newsMap:
#            self.__newsMap[title] = link

    def siteTitle(self):
        '''Returns the site title'''
        return self.__siteTitle

    def siteLink(self):
        '''Returns the site link'''
        return self.__siteLink

#    def setSiteTitle(self, title):
#        '''Sets the site\'s title'''
#        self.__siteTitle = title

#    def setSiteLink(self, link):
#        '''Sets the site\'s link'''
#        self.__siteLink = link

    def newsItems(self):
        '''Returns a map of all news items in the form map[title] == link'''
        return self.__newsMap

    

class TagHandler(ContentHandler):
    def __init__(self):
        self.__curTitle = ""
        self.__curLink = ""
        self.__curTag = ""
        self.siteLink = ""
        self.siteTitle = ""
        self.__isItem = 0
        self.itemList = []

    def startElement(self, name, attr):
        self.__curTag = name
        if name == "item":
            self.__isItem = 1
            
    def endElement(self, name):
        if name == "item":
            self.__isItem = 0
            self.__curLink = ""
            self.__curTitle = ""
        elif name == "channel":
            #if not len(self.itemList):
            #    print self.siteLink
            #    self.siteLink = self.__curLink
            #    self.siteTitle = self.__curTitle

            self.__curLink = ""
            self.__curTitle = ""    
        elif (name == 'link' or name == 'title') and \
                 (self.__curTitle.strip() and self.__curLink.strip()):
            # It's a link/title we want to ignore (for example <image>)
            if not self.__isItem:
                # we found our site title/link
                if not self.itemList:
                    self.siteTitle = self.__curTitle
                    self.siteLink = self.__curLink
                self.__curLink = self.__curTitle = ""
            else:
                self.itemList.append((self.__curTitle, self.__curLink))
                self.__curTitle = ""
                self.__curLink = ""
                self.__curTag = 'item'
            
    def characters(self, ch):
        if self.__curTag == 'title':
            if not self.siteTitle or self.__isItem:
                self.__curTitle += ch
        elif self.__curTag == 'link':
            if not self.siteLink or self.__isItem:
                self.__curLink += ch


if __name__ == '__main__':
    if len(sys.argv) > 1:
        f = open(sys.argv[1])
    	p = Parser(f)
    	print p.siteTitle(), ' link: ', p.siteLink(), '\n'
    	print p.newsItems()
    else:
        print 'Usage: \n\t ./parser.py file.rdf'
